akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowStatefulMapSpec.scala

```diff
@@ -0,0 +1,193 @@
+      val testSink = Source(List(1, 2, 3, 4, 5))
+        .statefulMap(() => 0)((agg, elem) => {
+          if (elem % 3 == 0)
+            throw ex
+          else
+            (agg + elem, (agg, elem))
+        }, _ => None)
+        .withAttributes(ActorAttributes.supervisionStrategy(Supervision.stoppingDecider))
+        .runWith(TestSink.probe[(Int, Int)])
+
+      testSink.expectSubscription().request(5)
+      testSink.expectNext((0, 1)).expectNext((1, 2)).expectError(ex)
+    }
+
+    "fail on upstream failure" in {
+      val (testSource, testSink) = TestSource
+        .probe[Int]
+        .statefulMap(() => 0)((agg, elem) => {
+          (agg + elem, (agg, elem))
+        }, _ => None)
+        .toMat(TestSink.probe[(Int, Int)])(Keep.both)
+        .run()
+
+      testSink.expectSubscription().request(5)
+      testSource.sendNext(1)
+      testSink.expectNext((0, 1))
+      testSource.sendNext(2)
+      testSink.expectNext((1, 2))
+      testSource.sendNext(3)
+      testSink.expectNext((3, 3))
+      testSource.sendNext(4)
+      testSink.expectNext((6, 4))
+      testSource.sendError(ex)
+      testSink.expectError(ex)
+    }
+
+    "defer upstream failure and remember state" in {
+      val (testSource, testSink) = TestSource
+        .probe[Int]
+        .statefulMap(() => 0)((agg, elem) => { (agg + elem, (agg, elem)) }, (state: Int) => Some((state, -1)))
+        .toMat(TestSink.probe[(Int, Int)])(Keep.both)
+        .run()
+
+      testSink.expectSubscription().request(5)
+      testSource.sendNext(1)
+      testSink.expectNext((0, 1))
+      testSource.sendNext(2)
+      testSink.expectNext((1, 2))
+      testSource.sendNext(3)
+      testSink.expectNext((3, 3))
+      testSource.sendNext(4)
+      testSink.expectNext((6, 4))
+      testSource.sendError(ex)
+      testSink.expectNext((10, -1))
+      testSink.expectError(ex)
+    }
+
+    "cancel upstream when downstream fail" in {
+      val promise = Promise[Done]()
+      val testSource = TestSource
+        .probe[Int]
+        .statefulMap(() => 100)((agg, elem) => {
+          (agg + elem, (agg, elem))
+        }, (state: Int) => {
+          promise.complete(Success(Done))
+          Some((state, -1))
+        })
+        .toMat(Sink.cancelled)(Keep.left)
+        .run()
+      testSource.expectSubscription().expectCancellation()
+      Await.result(promise.future, 3.seconds) shouldBe Done
+    }
+
+    "call its onComplete callback on abrupt materializer termination" in {
+      @nowarn("msg=deprecated")
+      val mat = ActorMaterializer()
+      val promise = Promise[Done]()
+
+      val matVal = Source
+        .single(1)
+        .statefulMap(() => -1)((_, elem) => (elem, elem), _ => {
+          promise.complete(Success(Done))
+          None
+        })
+        .runWith(Sink.never)(mat)
+      mat.shutdown()
+      matVal.failed.futureValue shouldBe a[AbruptStageTerminationException]
+      Await.result(promise.future, 3.seconds) shouldBe Done
```

Comment:
always try to call `onComplete`

Reply:
I see how this makes sense for the resource closing use case, but we should be careful to document very clearly that `onComplete` will be called in scenarios when there is no downstream and that the final element will be dropped for such cases.

Reply:
Yes, will update the document this weekend